File: packages/client/src/types/workspace.ts
/**
 * Workspace types for multi-workspace document and chats management
 * Local implementation replacing shared package
 */

// Base RAG configuration
export interface BaseRagConfig {
    embedding_model?: string;
    embedding_dim?: number | null;
    retriever_type?: string;
    chunk_size?: number;
    chunk_overlap?: number;
    top_k?: number;
    rerank_enabled?: boolean;
    rerank_model?: string | null;
}

// Vector RAG configuration
export interface VectorRagConfig extends BaseRagConfig {
    embedding_algorithm?: string;
    chunking_algorithm?: string;
    rerank_algorithm?: string;
}

// Graph RAG configuration
export interface GraphRagConfig extends BaseRagConfig {
    entity_extraction_algorithm?: string;
    relationship_extraction_algorithm?: string;
    clustering_algorithm?: string;
}

// Union type for RAG configurations
export type RagConfig = VectorRagConfig | GraphRagConfig;

// Workspace model
export interface Workspace {
    id: number;
    user_id?: number;
    name: string;
    description: string | null;
    rag_type: string;
    status?: string;
    created_at: string;
    updated_at: string;
    document_count?: number;
}

// Request types for workspace operations
export interface CreateWorkspaceRequest {
    name: string;
    description?: string;
    rag_type?: string;
    rag_config?: CreateRagConfigRequest;
}

export interface UpdateWorkspaceRequest {
    name?: string;
    description?: string;
}

// Request types for RAG config operations
export interface CreateRagConfigRequest {
    rag_type: string;
    config: RagConfig;
}

export interface UpdateRagConfigRequest {
    rag_type?: string;
    config?: RagConfig;
}

// Legacy type for backward compatibility - represents RAG config as flat object
export interface FlatRagConfig {
    embedding_algorithm?: string;
    chunking_algorithm?: string;
    rerank_algorithm?: string;
    chunk_size?: number;
    chunk_overlap?: number;
    top_k?: number;
    rerank_enabled?: boolean;
}

// Client-specific extensions
export interface WorkspaceState {
    workspaces: Workspace[];
    activeWorkspaceId: number | null;
    isLoading: boolean;
    error: string | null;
}

File: packages/server/src/infrastructure/models/workspace.py
"""Workspace domain models."""

from dataclasses import dataclass, field
from datetime import datetime


@dataclass
class Workspace:
    """Workspace model for organizing documents and chats sessions."""

    id: int
    user_id: int
    name: str
    description: str | None = None
    rag_type: str = "vector"  # "vector" or "graph"
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)

    def __repr__(self) -> str:
        return f"<Workspace(id={self.id}, name={self.name}, user_id={self.user_id})>"


@dataclass
class VectorRagConfig:
    """Vector RAG configuration for a workspace."""

    workspace_id: int
    embedding_algorithm: str = "ollama"
    chunking_algorithm: str = "sentence"
    rerank_algorithm: str = "none"
    chunk_size: int = 1000
    chunk_overlap: int = 200
    top_k: int = 5
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)


@dataclass
class GraphRagConfig:
    """Graph RAG configuration for a workspace."""

    workspace_id: int
    entity_extraction_algorithm: str = "spacy"
    relationship_extraction_algorithm: str = "dependency-parsing"
    clustering_algorithm: str = "leiden"
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)


# Backward compatibility - keep RagConfig for now
@dataclass
class RagConfig:
    """Generic RAG configuration (deprecated - use specific config types)."""

    workspace_id: int
    rag_type: str
    config: dict
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)